### DISCUSSION TITLE
exploring-dotnet-6-part-1-looking-inside-configurationmanager-in-dotnet-6/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#13] [Announcements]
# exploring-dotnet-6-part-1-looking-inside-configurationmanager-in-dotnet-6/

In this post I take a look at the ConfigurationManager class new in .NET 6, explore why it was added, and look at the scenarios it addresses.

https://andrewlock.net/exploring-dotnet-6-part-1-looking-inside-configurationmanager-in-dotnet-6/

[a comment by an automated account is not shown]

**Comment by nicoclau** [comment]
Hi Andrew, I noticed there were two different classes of ConfigurationManager

- Microsoft.Extensions.Configuration.ConfigurationManager used by WebApplicationBuilder available inside the nuget Microsoft.Extension.Configuration

- System.Configuration.ConfigurationManager defined in System.Configuration.ConfigurationManager.dll

When we use ASP.NET 6 we get in WebApplicationBuilder.Configuration Microsoft.Extensions.Configuration.ConfigurationManager

Why are there two classes?
How do we get back the AppSettings ?

In System.Configuration.ConfigurationManager we used to have

   public static NameValueCollection AppSettings
        {
            get
            {
                object section = GetSection("appSettings");
                if (!(section is NameValueCollection))
                {
                    // If config is null or not the type we expect, the declaration was changed.
                    // Treat it as a configuration error.
                    throw new ConfigurationErrorsException(SR.Config_appsettings_declaration_invalid);
                }

                return (NameValueCollection)section;
            }
        }

Now I need to do :

configuration.GetSection("xxx")["yy"]

to get from the appsetting.json:

"xxx":
{ "yy": "ww" }

It would be a good idea to update the article to help us understand better why and how to use the ConfigurationManager :) Thanks

**Comment by pkaushik-bee** [comment]
in line "To work around this, ConfigurationManager uses a custom IList<> implementation", seems like you were supposed to write "To work around this, ConfigurationSources uses a custom IList<> implementation"

**Reply by nicoclau** [reply]
yes it should be fixed

**Comment by GeorgeGomez** [comment]
Hello, and thanks for the very informative article. Question: what is the best way to add a json file provider and then conditionally add a KeyPerFile provider? I am currently adding the json file provider, building the configuration, reading a setting from the json file and finally loading up the KeyPerFile provider if the json file setting is set to true.

Is there a better way? 

Thanks in advance!
